// _@serverless-devs_dk-deploy-common@0.0.4-alpha.38@@serverless-devs/dk-deploy-common/lib/generateMiddlewares.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var parser = __importStar(require("@babel/parser"));
var t = __importStar(require("@babel/types"));
var traverse_1 = __importDefault(require("@babel/traverse"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var core_1 = require("@serverless-devs/core");
var lodash_1 = require("lodash");
var middlewares = [
    {
        name: 'cookiesMiddleware',
        policies: ['AliyunECSNetworkInterfaceManagementAccess', 'AliyunOTSFullAccess'],
    },
    {
        name: 'cookieSessionMiddleware',
        policies: ['AliyunECSNetworkInterfaceManagementAccess', 'AliyunOTSFullAccess'],
    },
];
function generateMiddlewares(options) {
    return __awaiter(this, void 0, void 0, function () {
        var codeUri, filepath, indexPath, content, ast;
        return __generator(this, function (_a) {
            codeUri = options.codeUri;
            filepath = path_1.default.resolve(codeUri);
            indexPath = path_1.default.join(filepath, 'index.js');
            content = fs_extra_1.default.readFileSync(indexPath, 'utf8');
            ast = parser.parse(content);
            traverse_1.default(ast, {
                Program: function (_a) {
                    var _this = this;
                    var node = _a.node;
                    var body = node.body;
                    body.forEach(function (item) {
                        if (t.isExpressionStatement(item)) {
                            if (t.isCallExpression(item.expression)) {
                                if (t.isMemberExpression(item.expression.callee)) {
                                    if (t.isIdentifier(item.expression.callee.property)) {
                                        var _loop_1 = function (mid) {
                                            if (item.expression.callee.property.name === 'use') {
                                                item.expression.arguments.forEach(function (obj) { return __awaiter(_this, void 0, void 0, function () {
                                                    return __generator(this, function (_a) {
                                                        switch (_a.label) {
                                                            case 0:
                                                                if (!t.isCallExpression(obj)) return [3 /*break*/, 2];
                                                                if (!t.isIdentifier(obj.callee)) return [3 /*break*/, 2];
                                                                if (!(obj.callee.name === mid.name)) return [3 /*break*/, 2];
                                                                return [4 /*yield*/, generateRole(options, mid.policies)];
                                                            case 1:
                                                                _a.sent();
                                                                _a.label = 2;
                                                            case 2: return [2 /*return*/];
                                                        }
                                                    });
                                                }); });
                                            }
                                        };
                                        for (var _i = 0, middlewares_1 = middlewares; _i < middlewares_1.length; _i++) {
                                            var mid = middlewares_1[_i];
                                            _loop_1(mid);
                                        }
                                    }
                                }
                            }
                        }
                    });
                },
            });
            return [2 /*return*/];
        });
    });
}
function generateRole(options, policies) {
    return __awaiter(this, void 0, void 0, function () {
        var app, templateFile, spath, sspath, content, appName, sprops, key;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    app = options.app;
                    if (!!lodash_1.get(app, 'role')) return [3 /*break*/, 3];
                    templateFile = process.env.templateFile;
                    spath = path_1.default.join(path_1.default.resolve(templateFile, '..'), '.s');
                    sspath = path_1.default.join(spath, path_1.default.basename(templateFile));
                    if (!fs_extra_1.default.existsSync(sspath)) {
                        fs_extra_1.default.copyFileSync(templateFile, sspath);
                    }
                    return [4 /*yield*/, core_1.getYamlContent(sspath)];
                case 1:
                    content = _a.sent();
                    appName = void 0;
                    sprops = void 0;
                    for (key in content.services) {
                        if (content.services[key].component.endsWith('jamstack-api')) {
                            appName = key;
                            sprops = content.services[key].props;
                        }
                    }
                    return [4 /*yield*/, core_1.modifyProps(appName, {
                            app: __assign(__assign({}, app), { role: {
                                    name: "fcdeploydefaultrole-" + lodash_1.get(app, 'name'),
                                    policies: getPolicies(sprops, policies),
                                } }),
                        }, sspath)];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
function getPolicies(sprops, policies) {
    var config = lodash_1.concat(lodash_1.get(sprops, 'app.role.policies', []), policies);
    return lodash_1.uniq(config);
}
module.exports = generateMiddlewares;
