// _@serverless-devs_dk-deploy-common@0.0.4-alpha.38@@serverless-devs/dk-deploy-common/lib/generateSwaggerUI.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var parser = __importStar(require("@babel/parser"));
var t = __importStar(require("@babel/types"));
var traverse_1 = __importDefault(require("@babel/traverse"));
var core_1 = require("@serverless-devs/core");
var path_1 = __importDefault(require("path"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var lodash_1 = require("lodash");
var logger = new core_1.Logger('dk-deploy-common');
function generateSwaggerUI(options) {
    return __awaiter(this, void 0, void 0, function () {
        var routes, _a, cwd, sourceCode, pathUrl, sourceCodePath, _i, routes_1, route, indexRoute, filepath, uiJson, paths, tags, dbJSON;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    logger.debug("\u51FD\u6570 generateSwaggerUI \u5165\u53C2: " + JSON.stringify(options, null, 2));
                    routes = options.routes, _a = options.cwd, cwd = _a === void 0 ? process.cwd() : _a, sourceCode = options.sourceCode, pathUrl = options.pathUrl;
                    sourceCodePath = path_1.default.join(cwd, '.s', sourceCode);
                    return [4 /*yield*/, ui({
                            sourceCode: sourceCode,
                            cwd: cwd,
                            pathUrl: pathUrl,
                        })];
                case 1:
                    _b.sent();
                    _i = 0, routes_1 = routes;
                    _b.label = 2;
                case 2:
                    if (!(_i < routes_1.length)) return [3 /*break*/, 5];
                    route = routes_1[_i];
                    indexRoute = route === '/' ? '/index' : route;
                    filepath = path_1.default.join(sourceCodePath, indexRoute);
                    return [4 /*yield*/, insertSwaggerUI({ filepath: filepath, indexRoute: indexRoute })];
                case 3:
                    uiJson = _b.sent();
                    if (!Object.keys(uiJson).length)
                        return [3 /*break*/, 4];
                    paths = lodash_1.get(uiJson, 'paths', {});
                    tags = lodash_1.get(uiJson, 'tags', []);
                    dbJSON = fs_extra_1.default.readJsonSync(path_1.default.join(sourceCodePath, 'ui/db.json'));
                    dbJSON.paths = Object.assign({}, dbJSON.paths, paths);
                    dbJSON.tags = __spreadArray(__spreadArray([], dbJSON.tags), tags);
                    fs_extra_1.default.outputJsonSync(path_1.default.join(sourceCodePath, 'ui/db.json'), dbJSON);
                    _b.label = 4;
                case 4:
                    _i++;
                    return [3 /*break*/, 2];
                case 5: return [2 /*return*/];
            }
        });
    });
}
// 修改 index.html, db.json
var ui = function (_a) {
    var sourceCode = _a.sourceCode, cwd = _a.cwd, pathUrl = _a.pathUrl;
    var sourceCodePath = path_1.default.join(cwd, '.s', sourceCode);
    var html = fs_extra_1.default.readFileSync(path_1.default.join(sourceCodePath, 'ui/index.html'), 'utf8');
    html = html.replace('https://petstore.swagger.io/v2/swagger.json', pathUrl + "/db.json");
    fs_extra_1.default.outputFileSync(path_1.default.join(sourceCodePath, 'ui/index.html'), html, 'utf8');
    var uiJson = fs_extra_1.default.readJsonSync(path_1.default.join(sourceCodePath, 'ui/db.json'));
    var _b = lodash_1.split(pathUrl, '://'), host = _b[1];
    uiJson.host = host;
    fs_extra_1.default.outputJsonSync(path_1.default.join(sourceCodePath, 'ui/db.json'), uiJson);
};
// 读取 index.js 下的 http api
var insertSwaggerUI = function (_a) {
    var filepath = _a.filepath, indexRoute = _a.indexRoute;
    return __awaiter(void 0, void 0, void 0, function () {
        var indexPath, content, ast, httpApiName, dkArguments, uiJson, apiObj;
        return __generator(this, function (_b) {
            logger.debug("\u51FD\u6570 insertSwaggerUI \u89E3\u6790\uFF1A" + filepath);
            indexPath = path_1.default.join(filepath, 'index.js');
            content = fs_extra_1.default.readFileSync(indexPath, 'utf8');
            ast = parser.parse(content);
            httpApiName = '';
            dkArguments = {};
            uiJson = {};
            traverse_1.default(ast, {
                CallExpression: function (_a) {
                    var node = _a.node;
                    if (t.isIdentifier(node.callee, { name: 'dk' }) && node.arguments && node.arguments.length) {
                        dkArguments = node.arguments[0];
                    }
                },
            });
            if (t.isIdentifier(dkArguments)) {
                httpApiName = dkArguments.name;
                if (!httpApiName)
                    return [2 /*return*/, uiJson];
                traverse_1.default(ast, {
                    VariableDeclarator: function (_a) {
                        var node = _a.node;
                        if (t.isIdentifier(node.id, { name: httpApiName }) && t.isObjectExpression(node.init)) {
                            var apiObj = node.init.properties;
                            uiJson = mapApi(apiObj, indexRoute);
                        }
                    },
                });
            }
            else if (t.isObjectExpression(dkArguments)) {
                apiObj = dkArguments.properties;
                uiJson = mapApi(apiObj, indexRoute);
            }
            logger.debug("\u51FD\u6570 insertSwaggerUI \u89E3\u6790\u7ED3\u679C\uFF1A" + JSON.stringify(uiJson));
            return [2 /*return*/, uiJson];
        });
    });
};
var mapApi = function (apiObj, indexRoute) {
    var paths = {};
    var tags = [];
    apiObj.forEach(function (api) {
        // 单行模式
        if (t.isObjectProperty(api) &&
            t.isStringLiteral(api.key) &&
            (t.isArrowFunctionExpression(api.value) || t.isFunctionExpression(api.value))) {
            var _a = lodash_1.split(api.key.value, ' '), apiMethod = _a[0], apiKey = _a[1];
            if (!apiMethod || !apiKey)
                return;
            if (!lodash_1.includes(tags, indexRoute)) {
                tags.push(indexRoute);
            }
            // const hasBody = includes(['post', 'patch', 'put'], apiMethod.toLowerCase());
            // let parameters = [];
            // if (hasBody) {
            //   parameters = [
            //     {
            //       "in": "body",
            //       "name": "body",
            //       "description": "body 数据",
            //       "required": false,
            //       "schema": {
            //         "$ref": "#/definitions/DefaultBody"
            //       }
            //     }
            //   ]
            // };
            lodash_1.set(paths, apiKey + "." + apiMethod.toLowerCase(), {
                tags: [indexRoute],
                operationId: apiMethod.toLowerCase() + "-" + apiKey,
                summary: '标题title',
                description: '描述',
                parameters: [],
                produces: ['application/json', 'application/xml'],
                responses: {},
            });
        }
        // 对象型
        if (t.isObjectProperty(api) && t.isStringLiteral(api.key) && t.isObjectExpression(api.value)) {
            var apiKey_1 = api.key.value;
            if (!lodash_1.includes(tags, indexRoute)) {
                tags.push(indexRoute);
            }
            var methodObj = api.value.properties;
            methodObj.forEach(function (method) {
                if (t.isObjectProperty(method) &&
                    t.isIdentifier(method.key) &&
                    (t.isArrowFunctionExpression(method.value) || t.isFunctionExpression(method.value))) {
                    var apiMethod = lodash_1.get(method.key, 'name');
                    // const hasBody = includes(['post', 'patch', 'put'], apiMethod.toLowerCase());
                    // let parameters = [];
                    // if (hasBody) {
                    //   parameters = [
                    //     {
                    //       "in": "body",
                    //       "name": "body",
                    //       "description": "body 数据",
                    //       "required": true,
                    //       "schema": {
                    //         "$ref": "#/definitions/DefaultBody"
                    //       }
                    //     }
                    //   ]
                    // };
                    lodash_1.set(paths, apiKey_1 + "." + apiMethod.toLowerCase(), {
                        tags: [indexRoute],
                        operationId: apiMethod.toLowerCase() + "-" + apiKey_1,
                        summary: '标题title',
                        description: '描述',
                        parameters: [],
                        produces: ['application/json', 'application/xml'],
                        responses: {},
                    });
                }
            });
        }
    });
    return { paths: paths, tags: tags };
};
module.exports = generateSwaggerUI;
