// _@serverless-devs_dk-deploy-common@0.0.4-alpha.38@@serverless-devs/dk-deploy-common/lib/generateTablestoreInitializer.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var parser = __importStar(require("@babel/parser"));
var t = __importStar(require("@babel/types"));
var traverse_1 = __importDefault(require("@babel/traverse"));
var generator_1 = __importDefault(require("@babel/generator"));
var prettier_1 = __importDefault(require("prettier"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var core_1 = require("@serverless-devs/core");
var lodash_1 = require("lodash");
var js_yaml_1 = __importDefault(require("js-yaml"));
var logger = new core_1.Logger('dk-deploy-common');
function generateTablestoreInitializer(options) {
    return __awaiter(this, void 0, void 0, function () {
        var templateFile, spath, filePath, _a, indexJs, configYml, indexPath, configPath;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    logger.debug("\u51FD\u6570 generateTablestoreInitializer \u5165\u53C2: " + JSON.stringify(options, null, 2));
                    templateFile = process.env.templateFile;
                    if (templateFile === 'null')
                        return [2 /*return*/];
                    spath = path_1.default.join(path_1.default.resolve(templateFile, '..'), '.s');
                    filePath = path_1.default.join(spath, options.sourceCode, path_1.default.basename(options.codeUri));
                    fs_extra_1.default.copySync(options.codeUri, filePath);
                    return [4 /*yield*/, insertTablestoreInitializer(options)];
                case 1:
                    _a = _b.sent(), indexJs = _a.indexJs, configYml = _a.configYml;
                    if (indexJs) {
                        indexPath = path_1.default.join(filePath, "index.js");
                        fs_extra_1.default.ensureFileSync(indexPath);
                        fs_extra_1.default.writeFileSync(indexPath, indexJs);
                    }
                    if (configYml) {
                        configPath = path_1.default.join(filePath, "config.yml");
                        fs_extra_1.default.ensureFileSync(configPath);
                        fs_extra_1.default.writeFileSync(configPath, js_yaml_1.default.dump(configYml));
                    }
                    return [2 /*return*/];
            }
        });
    });
}
// 修改index.js和config.yml内容，并返回
function insertTablestoreInitializer(options) {
    return __awaiter(this, void 0, void 0, function () {
        var codeUri, filepath, indexPath, content, ast, useTableStorePlugin, _a, useExportInitializer, initializerName, configYml, newCode, indexJs;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    codeUri = options.codeUri;
                    filepath = path_1.default.resolve(codeUri);
                    indexPath = path_1.default.join(filepath, 'index.js');
                    content = fs_extra_1.default.readFileSync(indexPath, 'utf8');
                    ast = parser.parse(content);
                    useTableStorePlugin = getUseTableStorePlugin(indexPath);
                    _a = useTableStore(ast), useExportInitializer = _a.useExportInitializer, initializerName = _a.initializerName;
                    logger.debug(codeUri + "/index.js\u6587\u4EF6\u662F\u5426\u4F7F\u7528tablestore: " + useTableStorePlugin);
                    logger.debug(codeUri + "/index.js\u6587\u4EF6\u662F\u5426\u5DF2\u7ECF\u5BFC\u51FA\u521D\u59CB\u5316\u51FD\u6570: " + useExportInitializer);
                    if (useExportInitializer) {
                        logger.debug(codeUri + "/index.js\u6587\u4EF6\u5BFC\u51FA\u521D\u59CB\u5316\u51FD\u6570\u540D\u79F0: " + initializerName);
                    }
                    if (!useTableStorePlugin)
                        return [2 /*return*/, { indexJs: content }];
                    return [4 /*yield*/, insertTablestoreinitializerYml({ filepath: filepath, initializerName: initializerName })];
                case 1:
                    configYml = _b.sent();
                    logger.debug(codeUri + "/config.yml: " + JSON.stringify(configYml, null, 2));
                    if (!useTableStorePlugin) return [3 /*break*/, 3];
                    return [4 /*yield*/, generateTablestoreRole(options)];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3:
                    // 如果使用tablestore但已经自定义initializer，直接返回原内容
                    if (useTableStorePlugin && useExportInitializer) {
                        return [2 /*return*/, { configYml: configYml, indexJs: content }];
                    }
                    traverse_1.default(ast, {
                        Program: function (_a) {
                            var node = _a.node;
                            var body = node.body;
                            var newImport = t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('exports'), t.identifier('initializer')), t.memberExpression(t.identifier('handler'), t.identifier('initializerHandler'))));
                            body.splice(body.length - 1, 0, newImport);
                        },
                    });
                    newCode = generator_1.default(ast).code;
                    indexJs = prettier_1.default.format(newCode, {
                        singleQuote: true,
                        trailingComma: 'es5',
                        printWidth: 100,
                    });
                    return [2 /*return*/, { configYml: configYml, indexJs: indexJs }];
            }
        });
    });
}
// 检查是否使用use tablestore
function getUseTableStorePlugin(indexPath) {
    var tablestoreRegx = /.use\(tablestoreInitialzerPlugin\(\)\)/g;
    var content = fs_extra_1.default.readFileSync(indexPath, 'utf8');
    return tablestoreRegx.test(content);
}
// 检查是否已经export.initializer
function useTableStore(ast) {
    var icount = 0;
    var initializerName;
    traverse_1.default(ast, {
        Program: function (_a) {
            var node = _a.node;
            var body = node.body;
            body.forEach(function (item) {
                // 是否已经export.initializer
                if (t.isExpressionStatement(item)) {
                    if (t.isAssignmentExpression(item.expression)) {
                        if (t.isMemberExpression(item.expression.right)) {
                            if (t.isIdentifier(item.expression.right.object)) {
                                item.expression.right.object.name === 'handler' && icount++;
                            }
                            if (t.isIdentifier(item.expression.right.property)) {
                                if (item.expression.right.property.name === 'initializerHandler') {
                                    icount++;
                                    if (t.isMemberExpression(item.expression.left)) {
                                        if (t.isIdentifier(item.expression.left.object)) {
                                            if (item.expression.left.object.name === 'exports') {
                                                icount++;
                                                if (t.isIdentifier(item.expression.left.property)) {
                                                    initializerName = item.expression.left.property.name;
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            });
        },
    });
    return {
        useExportInitializer: icount === 3,
        initializerName: initializerName,
    };
}
// config.yml配置初始化
function insertTablestoreinitializerYml(_a) {
    var filepath = _a.filepath, _b = _a.initializerName, initializerName = _b === void 0 ? 'initializer' : _b;
    return __awaiter(this, void 0, void 0, function () {
        var configPath, result;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    configPath = path_1.default.join(filepath, 'config.yml');
                    if (!fs_extra_1.default.existsSync(configPath)) return [3 /*break*/, 2];
                    return [4 /*yield*/, core_1.getYamlContent(configPath)];
                case 1:
                    result = _c.sent();
                    if (!lodash_1.get(result, 'function.initializer')) {
                        result.function.initializer = "index." + initializerName;
                    }
                    if (!lodash_1.get(result, 'function.initializationTimeout')) {
                        result.function.initializationTimeout = 60;
                    }
                    return [2 /*return*/, result];
                case 2: return [2 /*return*/, {
                        function: {
                            initializer: "index." + initializerName,
                            initializationTimeout: 60,
                        },
                    }];
            }
        });
    });
}
// 检测 s.yml 是否存在 role
function generateTablestoreRole(options) {
    return __awaiter(this, void 0, void 0, function () {
        var app, templateFile, spath, sspath, content, appName, key;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    app = options.app;
                    if (!!lodash_1.get(app, 'role')) return [3 /*break*/, 3];
                    templateFile = process.env.templateFile;
                    spath = path_1.default.join(path_1.default.resolve(templateFile, '..'), '.s');
                    sspath = path_1.default.join(spath, path_1.default.basename(templateFile));
                    fs_extra_1.default.copyFileSync(templateFile, sspath);
                    return [4 /*yield*/, core_1.getYamlContent(templateFile)];
                case 1:
                    content = _a.sent();
                    appName = void 0;
                    for (key in content.services) {
                        if (content.services[key].component.endsWith('jamstack-api')) {
                            appName = key;
                        }
                    }
                    return [4 /*yield*/, core_1.modifyProps(appName, {
                            app: __assign(__assign({}, app), { role: {
                                    name: "fcdeploydefaultrole-" + lodash_1.get(app, 'name'),
                                    policies: ['AliyunECSNetworkInterfaceManagementAccess', 'AliyunOTSFullAccess'],
                                } }),
                        }, sspath)];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
module.exports = generateTablestoreInitializer;
