// _@serverless-devs_dk@0.0.4-alpha.38@@serverless-devs/dk/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.github = exports.scheduler = exports.http = exports.oss = exports.cookieSessionMiddleware = exports.cookiesMiddleware = exports.validator = exports.mongodbInitialzerPlugin = exports.tablestoreInitialzerPlugin = exports.tablestoreEventParser = exports.dk = void 0;
var dk_1 = require("./dk");
Object.defineProperty(exports, "dk", { enumerable: true, get: function () { return __importDefault(dk_1).default; } });
var tablestore_event_parser_1 = require("@serverless-devs/tablestore-event-parser");
Object.defineProperty(exports, "tablestoreEventParser", { enumerable: true, get: function () { return __importDefault(tablestore_event_parser_1).default; } });
var tablestore_initialzer_plugin_1 = require("@serverless-devs/tablestore-initialzer-plugin");
Object.defineProperty(exports, "tablestoreInitialzerPlugin", { enumerable: true, get: function () { return __importDefault(tablestore_initialzer_plugin_1).default; } });
var mongodb_initialzer_plugin_1 = require("@serverless-devs/mongodb-initialzer-plugin");
Object.defineProperty(exports, "mongodbInitialzerPlugin", { enumerable: true, get: function () { return __importDefault(mongodb_initialzer_plugin_1).default; } });
var dk_validator_1 = require("@serverless-devs/dk-validator");
Object.defineProperty(exports, "validator", { enumerable: true, get: function () { return __importDefault(dk_validator_1).default; } });
var cookies_session_1 = require("@serverless-devs/cookies-session");
Object.defineProperty(exports, "cookiesMiddleware", { enumerable: true, get: function () { return cookies_session_1.cookiesMiddleware; } });
Object.defineProperty(exports, "cookieSessionMiddleware", { enumerable: true, get: function () { return cookies_session_1.cookieSessionMiddleware; } });
var oss_1 = require("./oss");
Object.defineProperty(exports, "oss", { enumerable: true, get: function () { return oss_1.oss; } });
var http_1 = require("./http");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return http_1.http; } });
var scheduler_1 = require("./scheduler");
Object.defineProperty(exports, "scheduler", { enumerable: true, get: function () { return scheduler_1.scheduler; } });
var github_1 = require("./github");
Object.defineProperty(exports, "github", { enumerable: true, get: function () { return github_1.github; } });
