// _@serverless-devs_tablestore-initialzer-plugin@0.0.4-alpha.38@@serverless-devs/tablestore-initialzer-plugin/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var tablestore_1 = __importDefault(require("tablestore"));
var tablestoreInitialzerPlugin = function () {
    var endpoint = process.env.tablestore_endpoint;
    var instanceName = process.env.tablestore_instanceName;
    return {
        initializer: function (_a) {
            var context = _a.context;
            var tableClient = new tablestore_1.default.Client({
                accessKeyId: context.credentials.accessKeyId,
                accessKeySecret: context.credentials.accessKeySecret,
                stsToken: context.credentials.securityToken,
                endpoint: endpoint,
                instancename: instanceName,
            });
            return {
                tableClient: tableClient,
                TableStore: tablestore_1.default,
            };
        },
    };
};
module.exports = tablestoreInitialzerPlugin;
