// _@serverless-devs_dk-core@0.0.4-alpha.38@@serverless-devs/dk-core/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var util_1 = require("./util");
var constant_1 = require("./constant");
var internal = {};
var dk = function (handler, baseMiddlewares) {
    var baseHandler;
    if (Object.prototype.toString.call(handler) === '[object Object]') {
        // 传入类型是对象
        baseHandler = function (config) { return util_1.httpRouteParserHandler(config, handler); };
    }
    else {
        baseHandler = handler || function () { };
    }
    var beforeMiddlewares = [];
    var afterMiddlewares = [];
    var onErrorMiddlewares = [];
    var initializerPlugins = [];
    var instance = function (first, second, thrid) {
        var _a = util_1.analizeEvents(first, second, thrid), type = _a.type, callback = _a.callback, event = _a.event, req = _a.req, res = _a.res, context = _a.context;
        var request = {
            req: req,
            res: res,
            event: event,
            context: context,
            callback: callback,
            type: type,
            result: undefined,
            error: undefined,
            internal: internal,
        };
        return runRequest(request, __spreadArray([], beforeMiddlewares), baseHandler, __spreadArray([], afterMiddlewares), __spreadArray([], onErrorMiddlewares));
    };
    instance.use = function (middlewares) {
        if (Array.isArray(middlewares)) {
            for (var _i = 0, middlewares_1 = middlewares; _i < middlewares_1.length; _i++) {
                var middleware = middlewares_1[_i];
                instance.applyMiddleware(middleware);
            }
            return instance;
        }
        return instance.applyMiddleware(middlewares);
    };
    instance.applyMiddleware = function (middleware) {
        var before = middleware.before, after = middleware.after, initializer = middleware.initializer, onError = middleware.onError;
        if (!before && !after && !onError && !initializer) {
            throw new Error('Middleware must be an object containing at least one key among "before", "after", "onError"');
        }
        if (before)
            instance.before(before);
        if (after)
            instance.after(after);
        if (initializer)
            instance.initializer(initializer);
        if (onError)
            instance.onError(onError);
        return instance;
    };
    instance.initializerHandler = undefined;
    instance.initializer = function (initializerPlugin) {
        initializerPlugins.push(initializerPlugin);
        // @ts-ignore
        instance.initializerHandler = fcInitializer(initializerPlugins);
        return instance;
    };
    // Inline Middlewares
    instance.before = function (beforeMiddleware) {
        beforeMiddlewares.push(beforeMiddleware);
        return instance;
    };
    instance.after = function (afterMiddleware) {
        afterMiddlewares.unshift(afterMiddleware);
        return instance;
    };
    instance.onError = function (onErrorMiddleware) {
        onErrorMiddlewares.push(onErrorMiddleware);
        return instance;
    };
    instance.__middlewares = {
        before: beforeMiddlewares,
        after: afterMiddlewares,
        onError: onErrorMiddlewares,
        initializer: initializerPlugins,
    };
    baseMiddlewares && instance.use(baseMiddlewares);
    return instance;
};
var runRequest = function (request, beforeMiddlewares, baseHandler, afterMiddlewares, onErrorMiddlewares) { return __awaiter(void 0, void 0, void 0, function () {
    var _a, e_1, e_2;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                _b.trys.push([0, 7, 12, 13]);
                if (!(request.type === 'HTTP')) return [3 /*break*/, 2];
                return [4 /*yield*/, util_1.getBody(request)];
            case 1:
                _b.sent();
                _b.label = 2;
            case 2: return [4 /*yield*/, runMiddlewares(request, beforeMiddlewares)];
            case 3:
                _b.sent();
                if (!(request.result === undefined)) return [3 /*break*/, 6];
                _a = request;
                return [4 /*yield*/, baseHandler(util_1.analizeRequestParams(request))];
            case 4:
                _a.result = _b.sent();
                return [4 /*yield*/, runMiddlewares(request, afterMiddlewares)];
            case 5:
                _b.sent();
                _b.label = 6;
            case 6: return [3 /*break*/, 13];
            case 7:
                e_1 = _b.sent();
                // Reset result changes made by after stack before error thrown
                request.result = undefined;
                request.error = e_1;
                _b.label = 8;
            case 8:
                _b.trys.push([8, 10, , 11]);
                return [4 /*yield*/, runMiddlewares(request, onErrorMiddlewares)];
            case 9:
                _b.sent();
                // Catch if onError stack hasn't handled the error
                if (request.type === constant_1.HTTP) {
                    if (request.result === undefined || request.res === undefined) {
                        request.result = {
                            statusCode: typeof e_1.code === 'number' ? e_1.code : 500,
                            body: { code: e_1.code, message: e_1.message },
                        };
                    }
                }
                else {
                    request.callback(request.error);
                }
                return [3 /*break*/, 11];
            case 10:
                e_2 = _b.sent();
                // Save error that wasn't handled
                e_2.originalError = request.error;
                request.error = e_2;
                if (request.type === constant_1.HTTP) {
                    throw request.error;
                }
                else {
                    request.callback(request.error);
                }
                return [3 /*break*/, 11];
            case 11: return [3 /*break*/, 13];
            case 12: return [7 /*endfinally*/];
            case 13:
                util_1.makeResult(request);
                if (request.type === constant_1.INITIALIZER) {
                    Object.assign(internal, request.result);
                }
                return [2 /*return*/, request.error ? undefined : request];
        }
    });
}); };
var runMiddlewares = function (request, middlewares) { return __awaiter(void 0, void 0, void 0, function () {
    var _i, middlewares_2, nextMiddleware, res;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                _i = 0, middlewares_2 = middlewares;
                _a.label = 1;
            case 1:
                if (!(_i < middlewares_2.length)) return [3 /*break*/, 4];
                nextMiddleware = middlewares_2[_i];
                return [4 /*yield*/, (nextMiddleware === null || nextMiddleware === void 0 ? void 0 : nextMiddleware(request))];
            case 2:
                res = _a.sent();
                // short circuit chaining and respond early
                if (res !== undefined) {
                    request.result = res;
                    return [2 /*return*/];
                }
                _a.label = 3;
            case 3:
                _i++;
                return [3 /*break*/, 1];
            case 4: return [2 /*return*/];
        }
    });
}); };
var fcInitializer = function (initializerList) {
    return dk(function (context) { return __awaiter(void 0, void 0, void 0, function () {
        var items, _i, initializerList_1, item, _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    items = {};
                    _i = 0, initializerList_1 = initializerList;
                    _d.label = 1;
                case 1:
                    if (!(_i < initializerList_1.length)) return [3 /*break*/, 4];
                    item = initializerList_1[_i];
                    _b = (_a = Object).assign;
                    _c = [items];
                    return [4 /*yield*/, item(context)];
                case 2:
                    _b.apply(_a, _c.concat([_d.sent()]));
                    _d.label = 3;
                case 3:
                    _i++;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/, items];
            }
        });
    }); });
};
module.exports = dk;
