// _@serverless-devs_dk-core@0.0.4-alpha.38@@serverless-devs/dk-core/lib/util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpRouteParserHandler = exports.getBody = exports.analizeRequestParams = exports.makeResult = exports.analizeEvents = void 0;
var constant_1 = require("./constant");
var dk_util_1 = require("@serverless-devs/dk-util");
var body_1 = __importDefault(require("body"));
var compile = require('path-to-regexp').match;
var makeHttpResponse = function (_a, httpResp) {
    var statusCode = _a.statusCode, headers = _a.headers, deleteHeaders = _a.deleteHeaders, body = _a.body;
    // statusCode
    if (statusCode !== undefined) {
        if (httpResp.setStatusCode) { // fc 的 res 提供 setStatusCode 方法，本地模拟 http 请求时不存在，需手动修改 statusCode
            httpResp.setStatusCode(statusCode);
        }
        else {
            httpResp.statusCode = statusCode;
        }
    }
    // headers
    if (!dk_util_1.isContainerEmpty(headers)) {
        Object.keys(headers).forEach(function (key) {
            httpResp.setHeader(key, headers[key]);
        });
    }
    // deleteHeader
    if (!dk_util_1.isContainerEmpty(deleteHeaders)) {
        deleteHeaders.forEach(function (key) {
            httpResp.deleteHeader(key);
        });
    }
    httpResp.send(body);
};
/**
 * @description 解析函数的事件
 * exports.initializer = (context, callback)  initializer触发
 * exports.handler = (req, resp, context) http触发
 * exports.handler = (event, context, callback) 事件触发
 * @param first
 * @param second
 * @param thrid
 */
var analizeEvents = function (first, second, thrid) {
    var type;
    var callback;
    var event;
    var req;
    var res;
    var context = {};
    if (typeof thrid === 'undefined') {
        // initializer
        type = constant_1.INITIALIZER;
        context = first;
        callback = second;
        return {
            context: context,
            callback: callback,
            type: type,
        };
    }
    else if (typeof thrid === 'function') {
        // event trigger
        type = constant_1.EVENT;
        callback = thrid;
        context = second;
        event = first; // typeof event = Buffer
        return {
            event: event,
            context: context,
            callback: callback,
            type: type,
        };
    }
    else {
        type = constant_1.HTTP;
        req = first;
        res = second;
        context = thrid;
        return {
            req: req,
            res: res,
            context: context,
            type: type,
        };
    }
};
exports.analizeEvents = analizeEvents;
var makeResult = function (_a) {
    var res = _a.res, type = _a.type, callback = _a.callback, result = _a.result, error = _a.error;
    if (type === constant_1.INITIALIZER || type === constant_1.EVENT) {
        error ? callback(error) : callback(null, result);
    }
    else {
        makeHttpResponse(result, res);
    }
};
exports.makeResult = makeResult;
var analizeRequestParams = function (_a) {
    var internal = _a.internal, res = _a.res, req = _a.req, result = _a.result, context = _a.context, event = _a.event, type = _a.type;
    if (type === constant_1.INITIALIZER) {
        return {
            context: context,
            internal: internal,
        };
    }
    else if (type === constant_1.EVENT) {
        return {
            event: event,
            context: context,
            internal: internal,
        };
    }
    else {
        var restResponse = dk_util_1.omit(res, ['statusCode', 'headers', 'deleteHeaders', 'body']);
        return {
            req: req,
            res: dk_util_1.isContainerEmpty(restResponse) ? result : Object.assign({}, restResponse, result || {}),
            context: context,
            internal: internal,
        };
    }
};
exports.analizeRequestParams = analizeRequestParams;
var getBody = function (request) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/, new Promise(function (resolve, reject) {
                if (!request.req.on) {
                    resolve(request);
                }
                try {
                    body_1.default(request.req, function (e, b) {
                        request.req.body = b;
                        resolve(request);
                    });
                }
                catch (e) {
                    reject(e);
                }
            })];
    });
}); };
exports.getBody = getBody;
/*eslint guard-for-in: 0*/
var matchFunc = function (path, option) { return compile(path, option); };
/**
 * 兼容 method 大小写
 * @param request
 * @param route
 */
var httpRouteParserHandler = function (request, route) { return __awaiter(void 0, void 0, void 0, function () {
    var _a, method, path, _b, _c, _i, i, type, _d, routeMethod, url, match, uri, data, match, uri, newRouteObj, j, data;
    return __generator(this, function (_e) {
        switch (_e.label) {
            case 0:
                _a = request.req, method = _a.method, path = _a.path;
                _b = [];
                for (_c in route)
                    _b.push(_c);
                _i = 0;
                _e.label = 1;
            case 1:
                if (!(_i < _b.length)) return [3 /*break*/, 7];
                i = _b[_i];
                type = Object.prototype.toString.call(route[i]);
                if (!(type === '[object Function]' || type === '[object AsyncFunction]')) return [3 /*break*/, 4];
                _d = i.split(' '), routeMethod = _d[0], url = _d[1];
                match = matchFunc(url);
                uri = path.split('?')[0];
                if (!(match(uri) && routeMethod.toLowerCase() === method.toLowerCase() && route[i])) return [3 /*break*/, 3];
                request.req.match = match(path);
                return [4 /*yield*/, route[i](request)];
            case 2:
                data = _e.sent();
                return [2 /*return*/, data];
            case 3: return [3 /*break*/, 6];
            case 4:
                if (!(type === '[object Object]')) return [3 /*break*/, 6];
                match = matchFunc(i);
                uri = path.split('?')[0];
                newRouteObj = {};
                for (j in route[i]) {
                    newRouteObj[j.toLowerCase()] = route[i][j];
                }
                ;
                if (!(match(uri) && route[i] && newRouteObj[method.toLowerCase()])) return [3 /*break*/, 6];
                request.req.match = match(path);
                return [4 /*yield*/, newRouteObj[method.toLowerCase()](request)];
            case 5:
                data = _e.sent();
                return [2 /*return*/, data];
            case 6:
                _i++;
                return [3 /*break*/, 1];
            case 7: return [2 /*return*/, { body: '' }];
        }
    });
}); };
exports.httpRouteParserHandler = httpRouteParserHandler;
