// _@serverless-devs_dk@0.0.4-alpha.38@@serverless-devs/dk/lib/util/generateConfig.js
"use strict";
/**
 * 生成config文件
 * config文件 按照规范是在和当前文件同级的地方。
 * - config.yml
 * - index.js
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateConfig = void 0;
/**
 *  config.yml 配置如下
 ```
 function:
 runtime: nodejs12
 # memory: 1152
 # timeout: 60
 # concurrency: 1
 oss:
 bucketName: my-bukect
 events:
 - oss:ObjectCreated:*
 - oss:ObjectRemoved:DeleteObject
 filter:
 Key:
 Prefix: source/
 Suffix: .png
 ```
 如果代码和配置文件同时，这时候配置文件会优先于代码。
 */
var fs_extra_1 = __importDefault(require("fs-extra"));
var js_yaml_1 = __importDefault(require("js-yaml"));
var path_1 = __importDefault(require("path"));
var generateConfig = function (type, args) {
    var filePath = path_1.default.resolve(process.cwd(), 'config.yml');
    fs_extra_1.default.ensureFileSync(filePath);
    var content = js_yaml_1.default.load(fs_extra_1.default.readFileSync(filePath, 'utf8')) || {};
    // oss配置
    if (type === 'oss') {
        var oss = args.oss;
        oss.filter = {
            Key: {
                Prefix: oss.filter.prefix,
                Suffix: oss.filter.suffix,
            },
        };
        content.oss = oss;
    }
    // http配置
    if (type === 'http') {
        content.http = args.http;
    }
    // scheduler配置
    if (type === 'scheduler') {
        content.scheduler = args.scheduler;
    }
    // 配置写入config.yml文件
    fs_extra_1.default.writeFileSync(filePath, js_yaml_1.default.dump(content));
};
exports.generateConfig = generateConfig;
