// _@serverless-devs_dk@0.0.4-alpha.38@@serverless-devs/dk/lib/util/githubHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGithubHandler = void 0;
// https://docs.github.com/cn/developers/webhooks-and-events/webhooks/about-webhooks
// https://github.com/rvagg/github-webhook-handler
// https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads
var events_1 = __importDefault(require("events"));
var crypto_1 = __importDefault(require("crypto"));
var qs_1 = require("qs");
function findHandler(url, arr) {
    if (!Array.isArray(arr)) {
        return arr;
    }
    var ret = arr[0];
    for (var i = 0; i < arr.length; i++) {
        if (url === arr[i].path) {
            ret = arr[i];
        }
    }
    return ret;
}
function checkType(options) {
    if (typeof options !== 'object') {
        throw new TypeError('must provide an options object');
    }
    var _a = options.secret, secret = _a === void 0 ? '' : _a, _b = options.path, path = _b === void 0 ? '' : _b;
    if (typeof path !== 'string') {
        throw new TypeError('must provide a \'path\' option');
    }
    if (typeof secret !== 'string') {
        throw new TypeError('must provide a \'secret\' option');
    }
}
// 生成令牌
function sign(data, secret) {
    if (secret === void 0) { secret = ''; }
    var hmac = crypto_1.default.createHmac('sha1', secret);
    hmac.update(Buffer.from(data));
    return "sha1=" + hmac.digest('hex');
}
// 校验令牌
function verify(signature, data, secret) {
    if (secret === void 0) { secret = ''; }
    if (!secret)
        return true;
    var sig = Buffer.from(signature);
    var signed = Buffer.from(sign(data, secret));
    if (sig.length !== signed.length) {
        return false;
    }
    return crypto_1.default.timingSafeEqual(sig, signed);
}
var createGithubHandler = function (initOptions) {
    var options;
    // validate type of options
    if (Array.isArray(initOptions)) {
        for (var i = 0; i < initOptions.length; i++) {
            checkType(initOptions[i]);
        }
    }
    else {
        checkType(initOptions);
    }
    var handler = function (req) {
        var events;
        options = findHandler(req.path, initOptions);
        if (typeof options.events === 'string' && options.events !== '*') {
            events = [options.events];
        }
        else if (Array.isArray(options.events) && options.events.indexOf('*') === -1) {
            events = options.events;
        }
        if (req.path !== options.path || req.method !== 'POST') {
            return { code: 404, message: 'The interface does not match github' };
        }
        function hasError(msg) {
            var err = new Error(msg);
            // 与最初设计违背，目前是直接执行，不是通过 emit 事件监听触发
            // handler.emit('err', err, req); //  handler.emit('error') 事件会导致整个流程直接中断抛异常，所以改成了handler.emit('err')
            return { code: 400, message: err.message };
        }
        var sig = req.headers['x-hub-signature'];
        var event = req.headers['x-github-event'];
        var id = req.headers['x-github-delivery'];
        var contentType = req.headers['content-type'];
        if (options.secret && !sig) {
            return hasError('No X-Hub-Signature found on request');
        }
        if (!event) {
            return hasError('No X-Github-Event found on request');
        }
        if (!id) {
            return hasError('No X-Github-Delivery found on request');
        }
        if (events && events.indexOf(event) === -1) {
            return hasError('X-Github-Event is not acceptable');
        }
        var body = req.body;
        var mimePattern = /^application\/x-www-form-urlencoded(;.*)?$/;
        // 令牌校验, 这里的 body 已经在 dk-core 中转为对象
        if (!verify(sig, mimePattern.test(contentType) ? qs_1.stringify(body) : JSON.stringify(body), options.secret)) {
            return hasError('X-Hub-Signature does not match blob signature');
        }
        var emitData = {
            event: event,
            id: id,
            payload: mimePattern.test(contentType) ? JSON.parse(body.payload) : body,
            protocol: req.protocol,
            host: req.headers.host,
            url: req.url,
            path: options.path
        };
        // 与最初设计违背，目前是直接执行，不是通过 emit 事件监听触发
        // handler.emit(event, emitData)
        // handler.emit('event', emitData)
        return { code: 200, message: 'success', data: emitData };
    };
    // make it an EventEmitter
    Object.setPrototypeOf(handler, events_1.default.prototype);
    events_1.default.call(handler);
    handler.sign = sign;
    handler.verify = verify;
    return handler;
};
exports.createGithubHandler = createGithubHandler;
