// _@serverless-devs_cookies-session@0.0.4-alpha.38@@serverless-devs/cookies-session/lib/tablestore/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
var tablestoreInitialzerPlugin = require('@serverless-devs/tablestore-initialzer-plugin');
var tableClientcache = {};
var tablestore = function (ctx, options) {
    var _a = tablestoreInitialzerPlugin().initializer({ context: ctx }), tableClient = _a.tableClient, TableStore = _a.TableStore;
    var tableName = options.name;
    var createTable = function () { return __awaiter(_this, void 0, void 0, function () {
        var params, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    params = {
                        tableMeta: {
                            tableName: tableName,
                            primaryKey: [
                                {
                                    name: 'uid',
                                    type: 'INTEGER'
                                }
                            ]
                        },
                        reservedThroughput: {
                            capacityUnit: {
                                read: 0,
                                write: 0
                            }
                        },
                        tableOptions: {
                            timeToLive: -1,
                            maxVersions: 1
                        }
                    };
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 5]);
                    return [4 /*yield*/, tableClient.describeTable({ tableName: tableName })];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 5];
                case 3:
                    error_1 = _a.sent();
                    return [4 /*yield*/, tableClient.createTable(params)];
                case 4:
                    _a.sent();
                    return [3 /*break*/, 5];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var putRow = function (data) {
        if (data === void 0) { data = {}; }
        return __awaiter(_this, void 0, void 0, function () {
            var attributeColumns, params, Data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        attributeColumns = Object.keys(data).map(function (key) {
                            var obj = {};
                            obj[key] = data[key];
                            return obj;
                        });
                        params = {
                            tableName: tableName,
                            condition: new TableStore.Condition(TableStore.RowExistenceExpectation.IGNORE, null),
                            primaryKey: [{ 'uid': TableStore.Long.fromNumber(20013) }],
                            attributeColumns: attributeColumns,
                            returnContent: { returnType: TableStore.ReturnType.Primarykey }
                        };
                        return [4 /*yield*/, tableClient.putRow(params)];
                    case 1:
                        Data = _a.sent();
                        return [2 /*return*/, Data];
                }
            });
        });
    };
    var GetRow = function () { return __awaiter(_this, void 0, void 0, function () {
        var params, Data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    params = {
                        tableName: tableName,
                        primaryKey: [{ 'uid': TableStore.Long.fromNumber(20013) }],
                        maxVersions: 2 //最多可读取的版本数，设置为2即代表最多可读取2个版本。
                    };
                    return [4 /*yield*/, tableClient.getRow(params)];
                case 1:
                    Data = _a.sent();
                    return [2 /*return*/, Data];
            }
        });
    }); };
    var UpdateRow = function () {
    };
    var DeleteRow = function () { return __awaiter(_this, void 0, void 0, function () {
        var params, Data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    params = {
                        tableName: tableName,
                        condition: new TableStore.Condition(TableStore.RowExistenceExpectation.IGNORE, null),
                        primaryKey: [{ 'uid': TableStore.Long.fromNumber(20013) }],
                    };
                    return [4 /*yield*/, tableClient.deleteRow(params)];
                case 1:
                    Data = _a.sent();
                    return [2 /*return*/, Data];
            }
        });
    }); };
    return {
        init: createTable,
        put: putRow,
        get: GetRow,
        update: UpdateRow,
        remove: DeleteRow
    };
};
module.exports = tablestore;
