// _@serverless-devs_dk-validator@0.0.4-alpha.38@@serverless-devs/dk-validator/lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var http_errors_1 = __importDefault(require("http-errors"));
var _2019_js_1 = __importDefault(require("ajv/dist/2019.js"));
var ajv_i18n_1 = __importDefault(require("ajv-i18n"));
var ajv_formats_1 = __importDefault(require("ajv-formats"));
var ajv_formats_draft2019_1 = __importDefault(require("ajv-formats-draft2019"));
var jsonSafeParse = require('@serverless-devs/dk-util').jsonSafeParse;
var ajv;
var ajvDefaults = {
    strict: true,
    coerceTypes: 'array',
    allErrors: true,
    useDefaults: 'empty',
    messages: false, // allow i18n
};
var defaults = {
    eventSchema: null,
    bodySchema: null,
    urlSchema: null,
    outputSchema: null,
    ajvOptions: {},
    ajvInstance: undefined,
    defaultLanguage: 'en',
};
var validatorMiddleware = function (opts) {
    if (opts === void 0) { opts = {}; }
    var _a = __assign(__assign({}, defaults), opts), eventSchema = _a.eventSchema, bodySchema = _a.bodySchema, urlSchema = _a.urlSchema, outputSchema = _a.outputSchema, ajvOptions = _a.ajvOptions, ajvInstance = _a.ajvInstance, defaultLanguage = _a.defaultLanguage;
    var useSchemaForAll = function (schemaConfig) {
        if (schemaConfig)
            return Object.prototype.hasOwnProperty.call(schemaConfig, 'type');
        return false;
    };
    // TODO：请求 url post/list/ 能访问对应路由 post/list
    var getCorrespondingvalidSchema = function (request, schemaConfig) {
        var _a = request.req, method = _a.method, path = _a.path;
        var validSchema = null;
        for (var _i = 0, _b = Object.entries(schemaConfig); _i < _b.length; _i++) {
            var _c = _b[_i], key = _c[0], schema = _c[1];
            if (key.toLowerCase().replace(/\s+/g, "") === method.toLowerCase() + path.toLowerCase()) {
                try {
                    validSchema = compile(schema, ajvOptions, ajvInstance);
                }
                catch (err) {
                    throw new http_errors_1.default.InternalServerError("There may be a problem with the schema format, you can refer to the detailed error information: " + err);
                }
            }
        }
        return validSchema;
    };
    var validEventSchema = useSchemaForAll(eventSchema) ? compile(eventSchema, ajvOptions, ajvInstance) : null;
    var validBodySchema = useSchemaForAll(bodySchema) ? compile(bodySchema, ajvOptions, ajvInstance) : null;
    var validUrlSchema = useSchemaForAll(urlSchema) ? compile(urlSchema, ajvOptions, ajvInstance) : null;
    var validOutputSchema = useSchemaForAll(outputSchema) ? compile(outputSchema, ajvOptions, ajvInstance) : null;
    var validatorMiddlewareBefore = function (request) { return __awaiter(void 0, void 0, void 0, function () {
        var event, validEventSchemaTemp, validate, valid, error, language, validBodySchemaTemp, validate, valid, error, language, validUrlSchemaTemp, validate, valid, error, language;
        return __generator(this, function (_a) {
            event = Buffer.isBuffer(request.event)
                ? jsonSafeParse(request.event.toString())
                : request.event;
            if (event) {
                validEventSchemaTemp = null;
                if (!validEventSchema) {
                    validEventSchemaTemp = getCorrespondingvalidSchema(request, eventSchema);
                }
                // 没有找到匹配的就不校验
                if (!validEventSchema && !validEventSchemaTemp)
                    return [2 /*return*/];
                validate = validEventSchema || validEventSchemaTemp;
                valid = validate(event);
                if (!valid) {
                    error = new http_errors_1.default.BadRequest('Event object failed validation');
                    language = chooseLanguage(event, defaultLanguage);
                    ajv_i18n_1.default[language](validate.errors);
                    error.details = validate.errors;
                    throw error;
                }
            }
            else {
                // http函数
                // body
                if (bodySchema) {
                    validBodySchemaTemp = null;
                    if (!validBodySchema) {
                        validBodySchemaTemp = getCorrespondingvalidSchema(request, bodySchema);
                    }
                    if (!validBodySchema && !validBodySchemaTemp)
                        return [2 /*return*/];
                    validate = validBodySchema || validBodySchemaTemp;
                    valid = validate(request.req.body);
                    if (!valid) {
                        error = new http_errors_1.default.BadRequest('Body object failed validation');
                        language = chooseLanguage(request.req.body, defaultLanguage);
                        ajv_i18n_1.default[language](validate.errors);
                        error.details = validate.errors;
                        throw error;
                    }
                }
                // path和queries
                if (urlSchema) {
                    validUrlSchemaTemp = null;
                    if (!validUrlSchema) {
                        validUrlSchemaTemp = getCorrespondingvalidSchema(request, urlSchema);
                    }
                    if (!validUrlSchema && !validUrlSchemaTemp)
                        return [2 /*return*/];
                    validate = validUrlSchema || validUrlSchemaTemp;
                    valid = validate({ path: request.req.path, queries: request.req.queries });
                    if (!valid) {
                        error = new http_errors_1.default.BadRequest('Url object failed validation');
                        language = chooseLanguage(request.req.queries, defaultLanguage);
                        ajv_i18n_1.default[language](validate.errors);
                        error.details = validate.errors;
                        throw error;
                    }
                }
            }
            return [2 /*return*/];
        });
    }); };
    var validatorMiddlewareAfter = function (request) { return __awaiter(void 0, void 0, void 0, function () {
        var validOutputSchemaTemp, validate, valid, error;
        return __generator(this, function (_a) {
            validOutputSchemaTemp = null;
            if (!validOutputSchema) {
                validOutputSchemaTemp = getCorrespondingvalidSchema(request, outputSchema);
            }
            if (!validOutputSchema && !validOutputSchemaTemp)
                return [2 /*return*/];
            validate = validOutputSchema || validOutputSchemaTemp;
            valid = validate(request.result);
            if (!valid) {
                error = new http_errors_1.default.InternalServerError('Response object failed validation');
                error.details = validate.errors;
                error.response = request.result;
                throw error;
            }
            return [2 /*return*/];
        });
    }); };
    var usedInputSchema = eventSchema || bodySchema || urlSchema;
    var usedOutputSchema = outputSchema;
    return {
        before: usedInputSchema ? validatorMiddlewareBefore : null,
        after: usedOutputSchema ? validatorMiddlewareAfter : null,
    };
};
// This is pulled out due to it's performance cost (50-100ms on cold start)
// Precompile your schema during a build step is recommended.
var compile = function (schema, ajvOptions, ajvInstance) {
    if (ajvInstance === void 0) { ajvInstance = null; }
    // Check if already compiled
    if (typeof schema === 'function' || !schema)
        return schema;
    var options = __assign(__assign({}, ajvDefaults), ajvOptions);
    if (!ajv) {
        ajv = ajvInstance !== null && ajvInstance !== void 0 ? ajvInstance : new _2019_js_1.default(options);
        ajv_formats_1.default(ajv);
        ajv_formats_draft2019_1.default(ajv);
    }
    return ajv.compile(schema);
};
/* in ajv-i18n Portuguese is represented as pt-BR */
var languageNormalizationMap = {
    pt: 'pt-BR',
    'pt-br': 'pt-BR',
    pt_BR: 'pt-BR',
    pt_br: 'pt-BR',
    zh: 'zh-TW',
    'zh-tw': 'zh-TW',
    zh_TW: 'zh-TW',
    zh_tw: 'zh-TW',
};
var normalizePreferredLanguage = function (lang) { var _a; return (_a = languageNormalizationMap[lang]) !== null && _a !== void 0 ? _a : lang; };
var availableLanguages = Object.keys(ajv_i18n_1.default);
var chooseLanguage = function (_a, defaultLanguage) {
    var preferredLanguage = _a.preferredLanguage;
    if (preferredLanguage) {
        var lang = normalizePreferredLanguage(preferredLanguage);
        if (availableLanguages.includes(lang)) {
            return lang;
        }
    }
    return defaultLanguage;
};
module.exports = validatorMiddleware;
